go/libkb/saltpack_dec.go

```diff
@@ -73,3 +87,67 @@ func SaltpackDecrypt(
 	}
 	return mki, nil
 }
+
+func peekTypeAndMakeDecoder(dearmored io.Reader, keyring naclKeyring) (io.Reader, *saltpack.MessageKeyInfo, saltpack.SigningPublicKey, saltpack.MessageType, error) {
+	// How much do we need to peek to get at the mode number?
```

Comment:
I think the Good and True way to do this would be to change our decoders to accept the header bytes as a slice, rather than insisting on reading them from the stream. Then we could decode the header twice: first into a very generic struct that ignores everything after the mode, and then second into the mode-specific struct based on what we read the first time. This peek method was kind of a hack to get the PR out the door without refactoring everything.